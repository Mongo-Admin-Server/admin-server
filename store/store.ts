import { combineReducers, configureStore } from "@reduxjs/toolkit";
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from "reduxjs-toolkit-persist";
import storage from "reduxjs-toolkit-persist/lib/storage";

import {
  EqualityFn,
  useDispatch as useDispatchBase,
  useSelector as useSelectorBase,
} from 'react-redux';

import databaseSlice from "@/domain/usecases/database-slice";
import collectionSlice from "@/domain/usecases/collection-slice";
import settingSlice from "@/domain/usecases/setting-slice";
import authSlice from "@/domain/usecases/auth-slice";
import documentSlice from "@/domain/usecases/document-slice";
import userSlice from "@/domain/usecases/user-slice";
import indexesSlice from "@/domain/usecases/indexes-slice";

export const reducer = combineReducers({
  database: databaseSlice,
  collection: collectionSlice,
  setting: settingSlice,
  auth: authSlice,
  document: documentSlice,
  user: userSlice,
  indexes: indexesSlice,
});

export const persistConfig = {
  key: 'root',
  storage,
};

const persistedReducer = persistReducer(persistConfig, reducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware({
    serializableCheck: false,
  }),
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export type AppStore = typeof store

export const useDispatch = () => useDispatchBase<AppDispatch>()

export const useSelector = <TSelected = unknown>(
  selector: (state: RootState) => TSelected,
  equalityFn?: EqualityFn<TSelected> | undefined
): TSelected => useSelectorBase<RootState, TSelected>(selector, equalityFn);
