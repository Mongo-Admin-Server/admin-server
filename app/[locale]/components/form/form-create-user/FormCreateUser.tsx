'use client';

import { useEffect, useState } from 'react';
import { useTranslations } from 'next-intl';

import GenericInput from '@components/ui/inputs/generic/GenericInput';
import GenericModal from '@components/modal/generic/GenericModal';
import SelectInput from '@components/ui/inputs/select/SelectInput';
import style from "./form-create-user.module.scss";

import { useDispatch, useSelector } from '@/store/store';
import { postUser, setErrorUser, selectError } from '@/domain/usecases/user-slice';
import { RoleType } from '@/domain/entities/user-types';

interface CreateUserModalProps {
  open: boolean;
  onClose: () => void;
}

export default function FormCreateUser({
  open,
  onClose,
}: CreateUserModalProps) {
  /* Static Data */
  const dispatch = useDispatch();
  const t = useTranslations();
  const errorSelector = useSelector(selectError);

  const optionsRoles = [
    { value: 'read', label: t('role.read'), },
    { value: 'readWrite', label: t('role.readWrite'), },
    { value: 'dbAdmin', label: t('role.dbAdmin'), },
    { value: 'dbOwner', label: t('role.dbOwner'), },
    { value: 'userAdmin', label: t('role.userAdmin'), },
    { value: 'clusterAdmin', label: t('role.clusterAdmin')},
    { value: 'backup', label: t('role.backup')},
    { value: 'root', label: t('role.root')}
  ];
  /* Methods */
  const handleSubmit = async () => {
    setRoles([{
      role,
      db: databaseName
    }])
    if (!errorInput && userName.trim() !== '') {
      await dispatch(postUser({
        username: userName,
        password: password,
        roles: roles
      })).then((result) => {
        if (result.meta.requestStatus === 'fulfilled') onClose();
      }); 
    }
  };
  
  const handleUserNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newUserName = event.target.value;
    setUserName(newUserName);

    if (newUserName.includes(' ')) {
      setErrorInput(t('formCreateCollection.spacesNotAllowedErrorMessage'))
    } else {
      setErrorInput('')
      dispatch(setErrorUser(''));
    }
  };

  /* Local Data */
  const [userName, setUserName] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [databaseName, setDatabaseName] = useState<string>('');
  const [role, setRole] = useState<string>('');
  const [roles, setRoles] = useState<RoleType[]>([]);
  const [errorInput, setErrorInput] = useState<string>('');
  const isSubmitButtonDisabled = errorInput !== '' || userName.trim() === '' || errorSelector !== '' ;
  const errors = errorInput || errorSelector;

  useEffect(() => {
    dispatch(setErrorUser(''));
    setUserName('');
    setPassword('');
    setDatabaseName('')
  }, [open, dispatch])
  
  return (
    <GenericModal
      open={open}
      title={t('user.create')}
      onClose={onClose}
      confirmLabelTrad={t('modal.button.create')}
      typeButton='submit'
      disabledButton={isSubmitButtonDisabled}
      onValidate={handleSubmit}
    >
      <form className={style['form-create-user']}>
        <GenericInput
          type="text"
          label={t('user.name')}
          value={userName}
          error={errors}
          placeholder={t('user.name')}
          onChange={handleUserNameChange}
        />
        <GenericInput
          type="password"
          label={t('user.password')}
          value={password}
          placeholder={t('user.password')}
          onChange={(event) => setPassword(event.target.value)}
        />
        <GenericInput
          type="text"
          label={t('user.database')}
          value={databaseName}
          placeholder={t('user.database')}
          onChange={(event) => setDatabaseName(event.target.value)}
        />
        <SelectInput
          label={t('user.role')}
          options={optionsRoles}
          value={role} 
          onChange={(event) => setRole(event.target.value)}
        />
      </form>
    </GenericModal>
  );
}
