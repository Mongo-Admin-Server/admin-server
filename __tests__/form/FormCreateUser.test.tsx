/* eslint-disable no-undef */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { NextIntlClientProvider } from 'next-intl';
import { store } from '@/store/store';
import { Provider } from 'react-redux';
import FormCreateUser from '@/app/[locale]/components/form/form-create-user/FormCreateUser';

jest.mock('next-intl/client', () => ({
  usePathname() {
    return '';
  },
  useRouter() {
    return {
      push: () => jest.fn(),
      replace: () => jest.fn(),
    };
  },
}));

const locale = 'fr';
  const message = require(`../../shared/locales/${locale}.json`)

  describe("FormCreateUser", () => {
    it('renders', () => {
      const handleClose = jest.fn();
      const { container } = render(
        <Provider store={store}>
          <NextIntlClientProvider locale={locale} messages={message}>
            <FormCreateUser 
              open={true}
              onClose={handleClose}
            />
          </NextIntlClientProvider>
        </Provider>
      );
      expect(container).toMatchSnapshot();
    })
  })