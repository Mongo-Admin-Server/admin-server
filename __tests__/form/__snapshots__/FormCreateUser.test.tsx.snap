// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCreateUser renders 1`] = `
<div>
  <dialog
    class="modal"
  >
    <section
      class="modal__header"
    >
      <div
        class="modal__header--icon"
      >
        <div
          class="undefined pointer-cursor"
        />
      </div>
      <h2>
        Créer un utilisateur
      </h2>
    </section>
    <section
      class="modal__container"
    >
      <form
        class="form-create-user"
      >
        <div
          class="input-container"
        >
          <label
            class="label"
          >
            Nom d'utilisateurs
          </label>
          <input
            class="input "
            placeholder="Nom d'utilisateurs"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-container"
        >
          <label
            class="label"
          >
            Mot de passe
          </label>
          <input
            class="input "
            placeholder="Mot de passe"
            type="password"
            value=""
          />
          <div
            class="iconContainer"
          >
            <div
              class="undefined "
            />
          </div>
        </div>
        <div
          class="input-container"
        >
          <label
            class="label"
          >
            user.database
          </label>
          <input
            class="input "
            placeholder="user.database"
            type="text"
            value=""
          />
        </div>
        <div
          class="container"
        >
          <label>
            Rôles
          </label>
          <select
            class="select "
          >
            <option
              disabled=""
              value=""
            >
              Choisir
            </option>
            <option
              value="read"
            >
              Lecture seule
            </option>
            <option
              value="readWrite"
            >
              Lecture et écriture
            </option>
            <option
              value="dbAdmin"
            >
              Administrateur de base de données
            </option>
            <option
              value="dbOwner"
            >
              Propriétaire de la base de données
            </option>
            <option
              value="userAdmin"
            >
              Administrateur d'utilisateur
            </option>
            <option
              value="clusterAdmin"
            >
              Administrateur de cluster
            </option>
            <option
              value="backup"
            >
              Sauvegarde
            </option>
            <option
              value="root"
            >
              Superutilisateur
            </option>
          </select>
          <div
            class="iconContainer"
          >
            <div
              class="undefined "
            />
          </div>
        </div>
      </form>
    </section>
    <section
      class="modal__footer"
    >
      <button
        class="button full outline center"
        type="button"
      >
        Annuler
      </button>
      <button
        class="button full center"
        disabled=""
        type="submit"
      >
        Créer
      </button>
    </section>
  </dialog>
</div>
`;
